internal/lsp/foldingrange.go

```diff
@@ -0,0 +1,155 @@
+package lsp
+
+import (
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+
+	"github.com/styrainc/regal/internal/lsp/opa/scanner"
+	"github.com/styrainc/regal/internal/lsp/opa/tokens"
+	"github.com/styrainc/regal/internal/lsp/types"
+)
+
+type stack []scanner.Position
+
+func (s stack) Push(p scanner.Position) stack {
+	return append(s, p)
+}
+
+func (s stack) Pop() (stack, scanner.Position) {
+	l := len(s)
+
+	return s[:l-1], s[l-1]
+}
+
+func TokenFoldingRanges(policy string) []types.FoldingRange {
+	scn, err := scanner.New(strings.NewReader(policy))
+	if err != nil {
+		panic(err)
+	}
+
+	var lastPosition scanner.Position
+
+	curlyBraceStack := stack{}
+	bracketStack := stack{}
+	parensStack := stack{}
+
+	foldingRanges := make([]types.FoldingRange, 0)
+
+	for {
+		token, position, _, errors := scn.Scan()
+
+		if token == tokens.EOF || len(errors) > 0 {
+			break
+		}
```

Comment:
This could also go into the switch, can't it?

Reply:
That went into and endless loop, presumably because the `break` breaks out of the `switch` and not the loop.